lib/Echidna/Async.hs

```diff
@@ -0,0 +1,46 @@
+module Echidna.Async where
+
+import Control.Concurrent (threadDelay, forkFinally)
+import Control.Monad (void, when)
+import Control.Monad.Reader (MonadReader, asks, ask)
+import Control.Monad.State.Strict (MonadState, gets)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Data.IORef (atomicModifyIORef', readIORef)
+import Data.Time (LocalTime)
+
+import Echidna.Types.Campaign (CampaignEvent, WorkerState(..))
+import Echidna.Types.Config (Env(..))
+import Echidna.Utility (getTimestamp)
+
+awaitThreads :: Env -> IO ()
+awaitThreads env = do
+  n <- readIORef env.numUnfinishedThreads
+  when (n > 0) $ threadDelay 10_000 >> awaitThreads env
```

Comment:
This works but I wonder if there is another concurrency primitive we could use to block instead of threadDelay. I think Chan would work here but maybe there is something simpler. We also have to think about exceptions in those threads so they don't fail silently.

This pattern of spawning short-running jobs and waiting for them to finish sounds like it should be a solved problem 🤔 Maybe something from here https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html ?